sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -975,6 +939,73 @@ object PushPredicateThroughAggregate extends Rule[LogicalPlan] with PredicateHel
       } else {
         filter
       }
+
+    case filter @ Filter(condition, u: Union) =>
+      val output = u.output
+      val newChildren = u.children.map { child =>
+        val attrMap: Map[Expression, Expression] = output.zip(child.output).toMap
+        val newCond = condition transform {
+          case e if attrMap.contains(e) => attrMap(e)
+        }
+        Filter(newCond, child)
+      }
+      Union(newChildren)
+
+    case filter @ Filter(condition, i: Intersect) =>
+      // Intersect could change the rows, so non-deterministic predicate can't be pushed down
+      val (pushDown, stayUp) = splitConjunctivePredicates(condition).partition { cond =>
+        cond.deterministic
+      }
+      if (pushDown.nonEmpty) {
+        val pushDownCond = pushDown.reduceLeft(And)
+        val output = i.output
+        val newChildren = i.children.map { child =>
+          val attrMap: Map[Expression, Expression] = output.zip(child.output).toMap
+          val newCond = pushDownCond transform {
+            case e if attrMap.contains(e) => attrMap(e)
+          }
+          Filter(newCond, child)
+        }
+        val newIntersect = i.withNewChildren(newChildren)
+        if (stayUp.nonEmpty) {
+          Filter(stayUp.reduceLeft(And), newIntersect)
+        } else {
+          newIntersect
+        }
+      } else {
+        filter
+      }
+
+    case filter @ Filter(condition, e @ Except(left, _)) =>
+      pushDownPredicate(filter, e, e.left) { pushDown =>
+        e.copy(left = Filter(pushDown, left))
+      }
+
+    case filter @ Filter(condition, u: UnaryNode) if u.expressions.forall(_.deterministic) =>
```

Comment:
For `UnaryNode`, I think we can also push down non-deterministic filter, as long as this operator doesn't re-order the input rows(do we have such an operator?)